### PR TITLE
8296515: RISC-V: Small refactoring for MaxReductionV/MinReductionV/AddReductionV node implementation

### PR DESCRIPTION
**Pull request by zifeihan**
HI,

The MaxReductionV, MinReductionV, AddReductionV nodes currently implemented by riscv rvv can be implemented by calling shared functions, and the T_BYTE and T_SHORT types in the MaxReductionV and MinReductionV node implementations can also be implemented in the same way as the T_INT type.

Please take a look and have some reviews. Thanks a lot.

## Testing:
- hotspot and jdk tier1 on unmatched board without new failures
- test/jdk/jdk/incubator/vector/* with fastdebug on qemu

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296515](https://bugs.openjdk.org/browse/JDK-8296515): RISC-V: Small refactoring for MaxReductionV/MinReductionV/AddReductionV node implementation

### Reviewers
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Author)
 * [Dingli Zhang](https://openjdk.org/census#dzhang) (@DingliZhang - Author)
 * [Yanhong Zhu](https://openjdk.org/census#yzhu) (@yhzhu20 - Author)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11036/head:pull/11036` \
`$ git checkout pull/11036`

Update a local copy of the PR: \
`$ git checkout pull/11036` \
`$ git pull https://git.openjdk.org/jdk pull/11036/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11036`

View PR using the GUI difftool: \
`$ git pr show -t 11036`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11036.diff">https://git.openjdk.org/jdk/pull/11036.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 1]
```diff
@@ -818,8 +818,8 @@ instruct reduce_andI(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             "vmv.x.s $dst, $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
-    __ reduce_operation($dst$$Register, as_VectorRegister($tmp$$reg),
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/riscv_v.ad:821]: Given the predicate `Matcher::vector_element_basic_type(n->in(2)) != T_LONG`, is there a risk it's going to match for types that don't fit the assert at https://github.com/openjdk/jdk/pull/11036/files#diff-35eb1d2f1e2f0514dd46bd7fbad49ff2c87703d5a3041a6433956df00a3fe6e6R1696 ?

It seems the same approach as `reduce_addI` should be taken with the following predicate:
```
  predicate(Matcher::vector_element_basic_type(n->in(2)) == T_BYTE || 
            Matcher::vector_element_basic_type(n->in(2)) == T_SHORT ||
            Matcher::vector_element_basic_type(n->in(2)) == T_INT);
```

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 2]
```diff
@@ -852,8 +852,8 @@ instruct reduce_orI(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
...
-                        $src1$$Register, as_VectorRegister($src2$$reg), bt, REDUCTION_OP::OR);
+    __ rvv_reduce_integral($dst$$Register, as_VectorRegister($tmp$$reg),
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/riscv_v.ad:855]: Same as L821

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 3]
```diff
@@ -886,8 +886,8 @@ instruct reduce_xorI(iRegINoSp dst, iRegIorL2I src1, vReg src2, vReg tmp) %{
             "vmv.x.s $dst, $tmp" %}
   ins_encode %{
     BasicType bt = Matcher::vector_element_basic_type(this, $src2);
-    __ reduce_operation($dst$$Register, as_VectorRegister($tmp$$reg),
-                        $src1$$Register, as_VectorRegister($src2$$reg), bt, REDUCTION_OP::XOR);
+    __ rvv_reduce_integral($dst$$Register, as_VectorRegister($tmp$$reg),
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/riscv_v.ad:889]: Same as L821

File: src/hotspot/cpu/riscv/riscv_v.ad (hunk 1 above)
Comment (zifeihan) [reply src/hotspot/cpu/riscv/riscv_v.ad:821]: > Given the predicate `Matcher::vector_element_basic_type(n->in(2)) != T_LONG`, is there a risk it's going to match for types that don't fit the assert at https://github.com/openjdk/jdk/pull/11036/files#diff-35eb1d2f1e2f0514dd46bd7fbad49ff2c87703d5a3041a6433956df00a3fe6e6R1696 ?
> 
> It seems the same approach as `reduce_addI` should be taken with the following predicate:
> 
> ```
>   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_BYTE || 
>             Matcher::vector_element_basic_type(n->in(2)) == T_SHORT ||
>             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
> ```

Thanks, Referring to aarch64, currently AndReductionV, OrReductionV, XorReductionV will only match T_BYTE, T_SHORT, T_INT, T_LONG types, and there should be no such risk at present. But for the sake of this uncertain risk, I will modify it later to use the same predicate as reduce_addI.

Comment (zifeihan) [reply src/hotspot/cpu/riscv/riscv_v.ad:821]: > Given the predicate `Matcher::vector_element_basic_type(n->in(2)) != T_LONG`, is there a risk it's going to match for types that don't fit the assert at https://github.com/openjdk/jdk/pull/11036/files#diff-35eb1d2f1e2f0514dd46bd7fbad49ff2c87703d5a3041a6433956df00a3fe6e6R1696 ?
> 
> It seems the same approach as `reduce_addI` should be taken with the following predicate:
> 
> ```
>   predicate(Matcher::vector_element_basic_type(n->in(2)) == T_BYTE || 
>             Matcher::vector_element_basic_type(n->in(2)) == T_SHORT ||
>             Matcher::vector_element_basic_type(n->in(2)) == T_INT);
> ```

Thanks. referring to aarch64 before, for unnecessary risks, now the same predicate is used as reduce_addI.

File: src/hotspot/cpu/riscv/riscv_v.ad (hunk 3 above)
Comment (zifeihan) [reply src/hotspot/cpu/riscv/riscv_v.ad:895]: > Same as [#11036 (comment)](https://github.com/openjdk/jdk/pull/11036#discussion_r1016358443)

Thanks, fixed

**Review by DingliZhang — APPROVED** [review]
LGTM, thanks!

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp [hunk 4]
```diff
@@ -1690,27 +1690,36 @@ bool C2_MacroAssembler::in_scratch_emit_size() {
...
       break;
-    case REDUCTION_OP::OR:
+    case Op_OrReductionV:
       vredor_vs(tmp, src2, tmp);
       break;
-    case REDUCTION_OP::XOR:
+    case Op_XorReductionV:
       vredxor_vs(tmp, src2, tmp);
       break;
+    case Op_MaxReductionV:
```
Comment (yhzhu20) [inline src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:1717]: We implemented Byte/Short MaxReductionV like this patch before. but there are 2 test cases that failed:
jdk/jdk/incubator/vector/ShortMaxVectorTests.java 
jdk/jdk/incubator/vector/ByteMaxVectorTests.java
The reason is that "make_reduction_input" will return INT_MAX for Byte & Short type.
Have you met this problem?

Comment (yhzhu20) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:1717]: > 

I see that this has been fixed in the previous PR 8271515.
Approved.

**Review by yhzhu20 — APPROVED** [review]

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp (hunk 4 above)
Comment (DingliZhang) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:1717]: Hi @yhzhu20, thanks for the review! I think this problem was solved in https://github.com/openjdk/jdk/pull/5873/commits/4e7f606030d67a4f284c375d9c8811f525ae97fd ：
```diff
@@ -1105,7 +1145,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
     case Op_MinReductionV:
       switch (bt) {
         case T_BYTE:
+          return gvn.makecon(TypeInt::make(max_jbyte));
         case T_SHORT:
+          return gvn.makecon(TypeInt::make(max_jshort));
         case T_INT:
           return gvn.makecon(TypeInt::MAX);
         case T_LONG:
@@ -1120,7 +1162,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
     case Op_MaxReductionV:
       switch (bt) {
         case T_BYTE:
+          return gvn.makecon(TypeInt::make(min_jbyte));
         case T_SHORT:
+          return gvn.makecon(TypeInt::make(min_jshort));
         case T_INT:
           return gvn.makecon(TypeInt::MIN);
         case T_LONG:
```
The Function "make_reduction_input" will return different types for Byte & Short than for int now.

Comment (yhzhu20) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:1717]: > Hi @yhzhu20, thanks for the review! I think this problem was solved in [4e7f606](https://github.com/openjdk/jdk/commit/4e7f606030d67a4f284c375d9c8811f525ae97fd) ：
> 
> ```diff
> @@ -1105,7 +1145,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
>      case Op_MinReductionV:
>        switch (bt) {
>          case T_BYTE:
> +          return gvn.makecon(TypeInt::make(max_jbyte));
>          case T_SHORT:
> +          return gvn.makecon(TypeInt::make(max_jshort));
>          case T_INT:
>            return gvn.makecon(TypeInt::MAX);
>          case T_LONG:
> @@ -1120,7 +1162,9 @@ Node* ReductionNode::make_reduction_input(PhaseGVN& gvn, int opc, BasicType bt)
>      case Op_MaxReductionV:
>        switch (bt) {
>          case T_BYTE:
> +          return gvn.makecon(TypeInt::make(min_jbyte));
>          case T_SHORT:
> +          return gvn.makecon(TypeInt::make(min_jshort));
>          case T_INT:
>            return gvn.makecon(TypeInt::MIN);
>          case T_LONG:
> ```
> 
> The function "make_reduction_input" will return different types for Byte & Short than for int now.

Understood. Thank you.

Comment (DingliZhang) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:1717]: > > 
> 
> I see that this has been fixed in the previous PR 8271515. Approved.

Sorry, I didn't refresh the page in time, so I saw your message too late and repeated the same message.

**Review by RealFYang — APPROVED** [review]
Nice refactoring work. Looks good.

**Comment by zifeihan** [comment]
Thanks for review. And all of  `test/jdk/jdk/incubator/vector/*` test case passed.
/integrate

**Pull request closed without merging**